File: modules/static/js/room/receipt.js
/*
 * IRCd Charybdis 5/Matrix
 *
 * Copyright (C) 2017 Charybdis Development Team
 * Copyright (C) 2017 Jason Volk (jason@zemos.net)
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice is present in all copies.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

'use strict';


/**************************************
 * Receipts
 *
 */

room.receipt = {};

room.receipt.__proto__ = function(room)
{
	if(!maybe(() => room.local.receipt))
		room.local.receipt = {};

	this.local = room.local.receipt;
};

room.receipt.send = async function(event_id, type = "m.read", opts = {})
{
	if(this.control.receipt_in_progress)
		return false;

	this.control.receipt_in_progress = event_id;
	let request = client.m.rooms.receipt.post(this.id, event_id, type, opts); try
	{
		let data = await request.response;
		return;
	}
	finally
	{
		this.control.receipt_in_progress = null;
	}
};

room.receipt.send.current = function()
{
	let current = this.timeline.current_event();
	if(!current)
		return false;

	if(!current.event_id)
		return false;

	if(mc.event.is_local(current))
		return false;

	if(current.type == "m.receipt")
		return false;

	if(maybe(() => current["m.read"][mc.my.mxid]))
		return false;

	if(maybe(() => current.origin_server_ts <= this.local.ts))
		return false;

	let receipts =
	{
		[current.event_id]:
		{
			"m.read":
			{
				[mc.my.mxid]:
				{
					ts: mc.date.now(),
				},
			},
		},
	};

	this.local.ts = mc.now();
	this.receipt.update(receipts);
	return this.receipt.send(current.event_id);
};

room.receipt.update = function(receipts)
{
	Object.each(receipts, (event_id, read) =>
	{
		mayif(() => this.timeline.get(event_id), (event) =>
		{
			this.receipt.update.event(event, read);
		});
	});
};

/**
 * merges updates to an object now attached to the event "m.read"
 * which explicitly specified so other keys in the receipt may
 * not clobber.
 */
room.receipt.update.event = function(event, read)
{
	Object.update(event,
	{
		"m.read": read["m.read"],
	});
};

/**
 * Sorts the receipts by timestamp so they can be viewed in the order of
 * reception.
 *
 * THIS FUNCTION MUST HAVE STABLE RESULTS for Angular and receipts with
 * the same timestamp must have some alternative ordering property.
 *
 * @returns Array of mxids which are keys for receipts ordered by ts
 */
room.receipt.sorted = function(receipts)
{
	return Object.keys(receipts).sort((a, b) =>
	{
		let tsa = receipts[a].ts;
		let tsb = receipts[b].ts;

		if(tsa === undefined && tsb === undefined)
			return 0;
		else if(tsa === undefined)
			return -1;
		else if(tsb === undefined)
			return 1;
		else
			return tsa < tsb?  -1:
			       tsa > tsb?   1:
			       a < b?      -1:
			       a > b?       1:
			                    0;
	});

	return ret;
};

File: matrix/receipt.cc
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.

decltype(ircd::m::receipt::log)
ircd::m::receipt::log
{
	"m.receipt"
};

ircd::m::event::id::buf
ircd::m::receipt::read(const m::room::id &room_id,
                       const m::user::id &user_id,
                       const m::event::id &event_id,
                       const json::object &options)
{
	const m::user::room user_room
	{
		user_id
	};

	const auto evid
	{
		send(user_room, user_id, "ircd.read", room_id,
		{
			{ "event_id",    event_id                                       },
			{ "ts",          options.get("ts", ircd::time<milliseconds>())  },
			{ "m.hidden",    options.get("m.hidden", false)                 },
		})
	};

	log::info
	{
		log, "%s read by %s in %s options:%s",
		string_view{event_id},
		string_view{user_id},
		string_view{room_id},
		string_view{options},
	};

	return evid;
}

bool
ircd::m::receipt::get(const m::room::id &room_id,
                      const m::user::id &user_id,
                      const m::event::id::closure &closure)
{
	const m::user::room user_room
	{
		user_id
	};

	const auto event_idx
	{
		user_room.get(std::nothrow, "ircd.read", room_id)
	};

	return m::get(std::nothrow, event_idx, "content", [&closure]
	(const json::object &content)
	{
		const json::string &event_id
		{
			content["event_id"]
		};

		closure(event_id);
	});
}


/// Does the user wish to not send receipts for events sent by its specific
/// sender?
bool
ircd::m::receipt::ignoring(const m::user &user,
                           const m::event::id &event_id)
{
	bool ret{false};
	m::get(std::nothrow, event_id, "sender", [&ret, &user]
	(const string_view &sender)
	{
		const m::user::room user_room{user};
		ret = user_room.has("ircd.read.ignore", sender);
	});

	return ret;
}

/// Does the user wish to not send receipts for events for this entire room?
bool
ircd::m::receipt::ignoring(const m::user &user,
                           const m::room::id &room_id)
{
	const m::user::room user_room{user};
	return user_room.has("ircd.read.ignore", room_id);
}

bool
ircd::m::receipt::freshest(const m::room::id &room_id,
                           const m::user::id &user_id,
                           const m::event::id &event_id)
try
{
	const m::user::room user_room
	{
		user_id
	};

	const m::event::idx event_idx
	{
		index(std::nothrow, event_id)
	};

	if(!event_idx)
		return true;

	const auto last_idx
	{
		user_room.get(std::nothrow, "ircd.read", room_id)
	};

	if(!last_idx)
		return true;

	if(last_idx < event_idx)
		return true;

	return false;
}
catch(const std::exception &e)
{
	log::derror
	{
		log, "Freshness of receipt in %s from %s for %s :%s",
		string_view{room_id},
		string_view{user_id},
		string_view{event_id},
		e.what()
	};

	return true;
}

bool
ircd::m::receipt::exists(const m::room::id &room_id,
                         const m::user::id &user_id,
                         const m::event::id &event_id)
{
	const m::user::room user_room
	{
		user_id
	};

	bool ret{false};
	user_room.get(std::nothrow, "ircd.read", room_id, [&ret, &event_id]
	(const m::event &event)
	{
		const auto &content
		{
			at<"content"_>(event)
		};

		ret = unquote(content.get("event_id")) == event_id;
	});

	return ret;
}
